### PR TITLE
Drop py27

### PR DESCRIPTION
**Pull request by simo5**
RIP python2

### CODE REVIEW

**Comment by simo5** [comment]
@tiran do you think there is any other python2isms I should get rid off here?

File: jwcrypto/jwa.py [hunk 1]
```diff
@@ -1104,6 +1101,7 @@ def signing_alg(cls, name):
         try:
             return cls.instantiate_alg(name, use='sig')
         except KeyError:
+            # pylint: disable=raise-missing-from
```
Comment (tiran) [inline jwcrypto/jwa.py:1104]: Python 3 does exception chaining. I recommend to add ``raise InvalidJWAAlgorithm(...) from None`` unless you want to show the inner exception to your users.

File: jwcrypto/jwk.py [hunk 2]
```diff
@@ -451,13 +447,13 @@ def _generate_OKP(self, params):
...
-        except KeyError:
+        except KeyError as e:
             raise InvalidJWKValue('"%s" is not a supported curve for the '
-                                  'OKP key type' % params['crv'])
+                                  'OKP key type' % params['crv']) from e
         self._import_pyca_pri_okp(key, **params)
 
     def _okp_curve_from_pyca_key(self, key):
-        for name, val in iteritems(_OKP_CURVES_TABLE):
+        for name, val in iter(_OKP_CURVES_TABLE.items()):
```
Comment (tiran) [inline jwcrypto/jwk.py:456]: ``items`` returns an iterator.

```suggestion
        for name, val in _OKP_CURVES_TABLE.items():
```

File: jwcrypto/jwk.py [hunk 3]
```diff
@@ -509,7 +505,7 @@ def import_key(self, **kwargs):
                 while name in names:
                     names.remove(name)
 
-        for name, val in iteritems(JWKValuesRegistry[kty]):
+        for name, val in iter(JWKValuesRegistry[kty].items()):
```
Comment (tiran) [inline jwcrypto/jwk.py:508]:
```suggestion
        for name, val in JWKValuesRegistry[kty].items():
```

File: jwcrypto/jwk.py [hunk 4]
```diff
@@ -950,7 +947,7 @@ def thumbprint(self, hashalg=hashes.SHA256()):
         """
 
         t = {'kty': self.get('kty')}
-        for name, val in iteritems(JWKValuesRegistry[t['kty']]):
+        for name, val in iter(JWKValuesRegistry[t['kty']].items()):
```
Comment (tiran) [inline jwcrypto/jwk.py:950]:
```suggestion
        for name, val in JWKValuesRegistry[t['kty']].items():
```

File: jwcrypto/jwk.py [hunk 5]
```diff
@@ -1013,7 +1010,7 @@ def __setitem__(self, item, value):
         super(JWK, self).__setitem__(item, value)
 
     def update(self, *args, **kwargs):
-        for k, v in iteritems(dict(*args, **kwargs)):
+        for k, v in iter(dict(*args, **kwargs).items()):
```
Comment (tiran) [inline jwcrypto/jwk.py:1013]:
```suggestion
        for k, v in dict(*args, **kwargs).items():
```

File: jwcrypto/jwk.py [hunk 6]
```diff
@@ -1055,6 +1052,7 @@ def __getattr__(self, item):
                         return self.get(item)
             raise KeyError
         except KeyError:
+            # pylint: disable=raise-missing-from
             raise AttributeError
```
Comment (tiran) [inline jwcrypto/jwk.py:1056]:
```suggestion
            raise AttributeError(key) from None
```

File: jwcrypto/jwk.py [hunk 7]
```diff
@@ -1066,6 +1064,7 @@ def __setattr__(self, item, value):
                     self.__setitem__(item, value)
             super(JWK, self).__setattr__(item, value)
         except KeyError:
+            # pylint: disable=raise-missing-from
             raise AttributeError
```
Comment (tiran) [inline jwcrypto/jwk.py:1068]:
```suggestion
            raise AttributeError(key) from None
```

File: jwcrypto/jwk.py [hunk 8]
```diff
@@ -1130,7 +1129,7 @@ def __setitem__(self, key, val):
             super(JWKSet, self).__setitem__(key, val)
 
     def update(self, *args, **kwargs):
-        for k, v in iteritems(dict(*args, **kwargs)):
+        for k, v in iter(dict(*args, **kwargs).items()):
```
Comment (tiran) [inline jwcrypto/jwk.py:1132]:
```suggestion
        for k, v in dict(*args, **kwargs).items():
```

File: jwcrypto/jwk.py [hunk 9]
```diff
@@ -1151,7 +1150,7 @@ def export(self, private_keys=True, as_dict=False):
                               a JSON object
         """
         exp_dict = dict()
-        for k, v in iteritems(self):
+        for k, v in iter(self.items()):
```
Comment (tiran) [inline jwcrypto/jwk.py:1153]:
```suggestion
        for k, v in self.items():
```

File: jwcrypto/jwk.py [hunk 10]
```diff
@@ -1203,7 +1202,7 @@ def get_key(self, kid):
 
     def __repr__(self):
         repr_dict = dict()
-        for k, v in iteritems(self):
+        for k, v in iter(self.items()):
```
Comment (tiran) [inline jwcrypto/jwk.py:1205]:
```suggestion
        for k, v in self.items():
```

File: jwcrypto/jws.py [hunk 11]
```diff
@@ -400,6 +400,7 @@ def deserialize(self, raw_jws, key=None, alg=None):
             except ValueError:
                 c = raw_jws.split('.')
                 if len(c) != 3:
+                    # pylint: disable=raise-missing-from
                     raise InvalidJWSObject('Unrecognized representation')
```
Comment (tiran) [inline jwcrypto/jws.py:404]:
```suggestion
                    # pylint: disable=raise-missing-from
                    raise InvalidJWSObject('Unrecognized representation') from None
```

**Review by tiran — CHANGES_REQUESTED** [review]

**Comment by simo5** [comment]
@tiran thanks for pointing out the issue with iter and that I could use 'from None'. I definitely want to use from None in those places.
I think I fixed all the points raised in the last force push.

**Pull request merged**